File: src/TidyTree/LinkedYList.ts
export class LinkedYList {
  index: number;
  pos_y: number;
  next?: LinkedYList;

  constructor(index: number, pos_y: number, node?: LinkedYList) {
    this.index = index;
    this.pos_y = pos_y;
    this.next = node;
  }

  bottom() {
    return this.pos_y;
  }

  pop() {
    return this.next;
  }

  update(index: number, pos_y: number) {
    let node: LinkedYList = this;

    while (node.pos_y <= pos_y) {
      if (node.next) {
        node = node.next;
      } else {
        return new LinkedYList(index, pos_y);
      }
    }

    new LinkedYList(index, pos_y, node);
  }
}

File: rust/src/layout/linked_y_list.rs
pub struct LinkedYList {
    pub index: usize,
    y: f32,
    next: Option<Box<LinkedYList>>,
}

impl LinkedYList {
    pub fn new(index: usize, pos_y: f32) -> Self {
        LinkedYList {
            index,
            y: pos_y,
            next: None,
        }
    }

    pub fn bottom(&self) -> f32 {
        self.y
    }

    pub fn update(self, index: usize, y: f32) -> Self {
        let mut list = self;

        while list.y <= y {
            if let Some(next) = list.next.take() {
                list = *next;
            } else {
                return LinkedYList {
                    index,
                    y,
                    next: None,
                };
            }
        }

        LinkedYList {
            index,
            y,
            next: Some(Box::new(list)),
        }
    }

    pub fn pop(mut self) -> Option<Self> {
        self.next.take().map(|next| *next)
    }
}
